### ISSUE TITLE
KubeClarity Runtime scan fails | SBOM not generated when only Trivy configured as the analyzers cum scanner

### ISSUE THREAD

**Issue by sprathod369** [openclarity/kubeclarity#481]
#### What happened:
Triggered a runtime in the target namespace but the scan completes with below errors

#### What you expected to happen:
The Trivy Analyzer should have successfully created an SBOM which should have been analyzed by the Trivy Scanner

#### How to reproduce it (as minimally and precisely as possible):
Install the latest v2.19.0 Helm version of KubeClarity. 
Trivy analyser enabled to true and reflected in the **analyzerList**. 
Trivy scanner enabled to true and reflected in the **scannerList**. 
Trigger a on-demand runtime scan against a target namespace names "test"

Note: 
1) If I change my KubeClarity configuration to use **"Syft" + "Grype"** combination, the runtime scan against the same target namespace "test" works without any errors in UI.
2) If I change my KubeClarity configuration to use **"Syft Trivy" + "Grype Trivy"** combination, the runtime scans against the same target namespace "test" works without any errors in UI.

This indicates that a **"Trivy" + "Trivy"** combination is not working as the **SBOM generation fails in this mode**. Errors seen in the UI below.

#### Are there any error messages in KubeClarity logs?
Following error seen in the UI for Runtime scan
Failed to scan image "sprathod369/dep-dev:portal-v1.0.1". Effected pods: dep-portal-585dddf599-ghgdf. 
```
Reasons: failed to analyze image: failed to run job manager: 
1 error occurred: * "trivy" job failed: failed to generate SBOM: image scan error: scan error: unable to initialize a scanner: 
unable to initialize a docker scanner: 3 errors occurred: 
* unable to inspect the image (docker.io/sprathod369/dep-dev@sha256:809c55a38de46a5ee818ffb68976b16744af8952f2abf72f0c51c4e4ccc78762): 
  Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? 
* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory 
* containerd socket not found: /run/containerd/containerd.sock 
```

#### Environment:
- Kubernetes version : v1.27.3
- Helm version : v3.11.1
- KubeClarity version: v2.19.0
- KubeClarity Helm Chart version: v2.19.0
- Cloud provider or hardware configuration: on-prem, Ubuntu 20x running Kind

**Comment by FrimIdan** [comment]
Hi @sprathod369 I see that the image is private, right?
Is there an imagePullSecret defined for the pod? If so, can you share the content (excluding secrets)? I want to see the list of `auths`

**Comment by sprathod369** [comment]
Yes @FrimIdan. You are correct, the image is pulled from a private repository in DockerHub. I did not define any imagePullSecret in the pod definitions but did the same in the default service accounts of my kubeclarity and test namespaces as per below:
1) Ensure docker login works from the VM, K8s-Kind is running on
2) Create secrets named "basic-regcred" and patch SA in kubeclarity namespace so that pull from docker hub org is successfull
```
~# kubectl create secret generic basic-regcred --from-file=.dockerconfigjson=/root/.docker/config.json --type=kubernetes.io/dockerconfigjson --namespace=kubeclarity
secret/basic-regcred created
~# kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "basic-regcred"}]}' -n kubeclarity
serviceaccount/default patched
```

3) The pods and services used for testing are running in a "test" namespace so before deploying them, I did the same thing as point 2 in the "test" namespace as well 

As mentioned above "Syft" configuration in my "analyzerList" can successfully create a SBOM for my target running pods. Everything being the same, if I change my KubeClarity configuration to "Trivy" in the "analyzerList", it fails to generate a SBOM due to the above errors.  

Does this info help? Thanks for reverting back.

**Comment by FrimIdan** [comment]
That info helps, I think.
I would assume that you have more than single `auth` in your `/root/.docker/config.json` file, right?
Currently if you look at https://github.com/openclarity/kubeclarity/blob/main/shared/pkg/utils/trivy/registry.go#L26-L48, since trivy supports only one auth we needed to take the first entry, and it might not the right fit for the specific image.

Can you confirm you have more than a single `auth` in the docker config? If so, can you try to create a new secret with a single auth related to the `sprathod369` repo?

**Comment by sprathod369** [comment]
Yes, there's only single auth in the docker config. This is how it looks
```
~# cat /root/.docker/config.json
{
        "auths": {
                "https://index.docker.io/v1/": {
                        "auth": "xxxxxxxxxxxxxxxxx"
                }
        }
}
```

**Comment by FrimIdan** [comment]
@Tehsmash any idea what could be the issue here? trivy analyzer failed to analyze private image while syft analyzer succeeded with the same secret configuration.

**Comment by Tehsmash** [comment]
So when I run trivy directly against a private repo without any authentication I get one more error than we're seeing in the output above:

```
$ docker run aquasec/trivy:0.41.0 image --format cyclonedx docker.io/tehsmash/private
2023-08-29T07:16:44.963Z        INFO    "--format cyclonedx" disables security scanning. Specify "--scanners vuln" explicitly if you want to include vulnerabilities in the CycloneDX report.
2023-08-29T07:16:46.850Z        FATAL   image scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
        * unable to inspect the image (docker.io/tehsmash/private): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
        * unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
        * containerd socket not found: /run/containerd/containerd.sock
        * GET https://index.docker.io/v2/tehsmash/private/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:tehsmash/private Type:repository]]
```

Notice the "GET ...." error in addition to the ones attempting to talk to the container runtimes. From that I guess Trivy is either successfully querying docker.io and not erroring, or isn't even attempting it, most likely the later because otherwise it should be successful. I wonder if there is a missing configuration somewhere on our side.

**Comment by FrimIdan** [comment]
Running kubeclarity cli also shows 4 errors

```
$ make cli && ANALYZER_LIST=trivy ./cli/bin/cli analyze docker.io/tehsmash/private -i image -o test.json
Building CLI ...
total 344640
-rwxr-xr-x@ 1 idanf  staff  176451602 Aug 29 10:33 cli
INFO[0000] Called trivy analyzer on source image docker.io/tehsmash/private  analyzer=trivy app=kubeclarity
ERRO[0003] failed to generate SBOM: image scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
	* unable to inspect the image (docker.io/tehsmash/private): Error response from daemon: No such image: tehsmash/private:latest
	* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
	* containerd socket not found: /run/containerd/containerd.sock
	* GET https://index.docker.io/v2/tehsmash/private/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:tehsmash/private Type:repository]]
  analyzer=trivy app=kubeclarity
WARN[0003] "trivy" job failed: failed to generate SBOM: image scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
	* unable to inspect the image (docker.io/tehsmash/private): Error response from daemon: No such image: tehsmash/private:latest
	* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
	* containerd socket not found: /run/containerd/containerd.sock
	* GET https://index.docker.io/v2/tehsmash/private/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:tehsmash/private Type:repository]]
  app=kubeclarity
FATA[0003] Failed to run job manager: 1 error occurred:
	* "trivy" job failed: failed to generate SBOM: image scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
	* unable to inspect the image (docker.io/tehsmash/private): Error response from daemon: No such image: tehsmash/private:latest
	* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
	* containerd socket not found: /run/containerd/containerd.sock
	* GET https://index.docker.io/v2/tehsmash/private/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:tehsmash/private Type:repository]]


  app=kubeclarity
```

According to the code in [here](https://github.com/aquasecurity/trivy/blob/v0.41.0/pkg/fanal/image/image.go#L47-L109) the remote option is always on - that is very strange.

**Comment by sprathod369** [comment]
@FrimIdan, @Tehsmash - Just wondering if Trivy configuration in KubeClarity runtime scan works fine for you as part of generating an SBOM for **some public repo**? 
I briefly attempted a runtime scan from Kubeclarity UI against the public **webgoat/webgoat-8.0** image running in my cluster in a test namespace but it did not work either.

Can you suggest me a public repo that I can test and confirm if only Trivy configured as the analyzer works fine so that it's ruled out that the current issue where **Trivy fails to generate the SBOM in KubeClarity runtime scan is not specific with private repos?**

**Comment by FrimIdan** [comment]
@sprathod369 I see that `webgoat/webgoat-8.0` is a JAVA image so probably https://github.com/openclarity/kubeclarity/issues/488 is related.

**Comment by sprathod369** [comment]
Yes @FrimIdan - I get the same errors mentioned by you in the above reference ticket 488. In the meantime, let me try running the trivy analyzer against some other tech stack and a public repo to confirm if SBOM gets generated with this Kubeclarity configuration

**Comment by FrimIdan** [comment]
@sprathod369 we just released a new version v2.21.1 where you can find a fix for https://github.com/openclarity/kubeclarity/issues/488

Can you try again your scenario and see if this issue can be resolved?

**Comment by sprathod369** [comment]
@FrimIdan - Yes, I validated the runtime scan a moment back and it works like a charm with latest release v2.21.1. 
This is what I did
**Test 1** - Configured analyzerList = "trivy" and scannerList = "trivy", enabled Trivy server, disabled Grype server, deployed webgoat/webgoat-8.0 and triggered the runtime scan from the UI. The scan was successful without any issues.

**Test 2** - Configured analyzerList = "syft trivy" and scannerList = "grype trivy", enabled Trivy and Grype servers, deployed webgoat/webgoat-8.0 and triggered the runtime scan from the UI. The scan was successful without any issues.

![image](https://github.com/openclarity/kubeclarity/assets/73107830/9191fe57-f880-419d-9911-21e5d86baa28)

Thanks for your support and discussion - appreciate your inputs!!
